delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala

```diff
@@ -0,0 +1,144 @@
+import ch.epfl.bluebrain.nexus.delta.plugins.elasticsearch.model.ElasticSearchViewValue._
+import ch.epfl.bluebrain.nexus.delta.plugins.elasticsearch.model._
+import ch.epfl.bluebrain.nexus.delta.rdf.IriOrBNode.Iri
+import ch.epfl.bluebrain.nexus.delta.sdk.model.permissions.Permission
+import ch.epfl.bluebrain.nexus.delta.sdk.model.projects.ProjectRef
+import ch.epfl.bluebrain.nexus.delta.sdk.utils.{IOUtils, UUIDF}
+import io.circe.Json
+import monix.bio.{IO, UIO}
+
+class ElasticSearchViews {}
+
+object ElasticSearchViews {
+
+  /**
+    * The elasticsearch module type.
+    */
+  final val moduleType: String = "elasticsearchview"
+
+  private[elasticsearch] def next(
+      state: ElasticSearchViewState,
+      event: ElasticSearchViewEvent
+  ): ElasticSearchViewState = {
+    // format: off
+    def created(e: ElasticSearchViewCreated): ElasticSearchViewState = state match {
+      case Initial     => Current(e.id, e.project, e.uuid, e.value, e.source, Map.empty, e.rev, deprecated = false,  e.instant, e.subject, e.instant, e.subject)
+      case s: Current  => s
+    }
+
+    def updated(e: ElasticSearchViewUpdated): ElasticSearchViewState = state match {
+      case Initial    => Initial
+      case s: Current => s.copy(rev = e.rev, value = e.value, source = e.source, updatedAt = e.instant, updatedBy = e.subject)
+    }
+
+    def tagAdded(e: ElasticSearchViewTagAdded): ElasticSearchViewState = state match {
+      case Initial    => Initial
+      case s: Current => s.copy(rev = e.rev, tags = s.tags + (e.tag -> e.targetRev), updatedAt = e.instant, updatedBy = e.subject)
+    }
+    // format: on
+
+    def deprecated(e: ElasticSearchViewDeprecated): ElasticSearchViewState = state match {
+      case Initial    => Initial
+      case s: Current => s.copy(rev = e.rev, deprecated = true, updatedAt = e.instant, updatedBy = e.subject)
+    }
+
+    event match {
+      case e: ElasticSearchViewCreated    => created(e)
+      case e: ElasticSearchViewUpdated    => updated(e)
+      case e: ElasticSearchViewTagAdded   => tagAdded(e)
+      case e: ElasticSearchViewDeprecated => deprecated(e)
+    }
+  }
+
+  private[elasticsearch] def evaluate(
+      validatePermission: Permission => IO[PermissionIsNotDefined, Unit],
+      validateMapping: Json => IO[InvalidElasticSearchMapping, Unit],
+      validateRef: ViewRef => IO[InvalidViewReference, Unit]
+  )(state: ElasticSearchViewState, cmd: ElasticSearchViewCommand)(implicit
+      clock: Clock[UIO] = IO.clock,
+      uuidF: UUIDF = UUIDF.random
+  ): IO[ElasticSearchViewRejection, ElasticSearchViewEvent] = {
+
+    def validate(id: Iri, project: ProjectRef, value: ElasticSearchViewValue): IO[ElasticSearchViewRejection, Unit] =
+      value match {
+        case v: AggregateElasticSearchViewValue =>
+          val ref = ViewRef(project, id)
+          for {
+            _ <- if (v.views.contains(ref)) IO.raiseError(CannotSelfReference(ref)) else IO.unit
```

Comment:
So far an `AggregatedElasticSearchView` can only reference to `ElasticSearchView`(s), not to other `AggregatedElasticSearchView`(s). Being that the case, this extra check is irrelevant, since `validateRef` function should already be checking that.

Reply:
I was thinking that we could allow AggregateElasticSearchViews in references and flatten their references. Happy to remove it if you don't find it useful.

Reply:
If we introduce this, we also introduce the problem of circular references so we could let that for later no ?

Reply:
we only need to check direct references because they're most likely a mistake; for the rest it can be filtered out.

Reply:
It could potentially be useful (but I'm not aware of a current use case for it). It would also be a bit more complicated to implement. I don't have any strong opinion, so as you prefer.

In case you go for it, please add this to the documentation additions.

Reply:
yeah, I'll remove it for now.